### PR TITLE
Add norwegian bokmål ('nb') lemmatizer and tag_map

### PR DESCRIPTION
**Pull request by ohenrik**
## Description
Adding support for lemmatization and simple tagging using tag_map

### Types of change
New features to language nb (Norwegian bokmål)

## Checklist
**Will check if boxes bellow as i get closer to completing. Need to start the PR to get help with tag_map. Do not merge yet!** 
- [x] I have submitted the spaCy Contributor Agreement.
- [ ] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by honnibal** [comment]
Thanks, looks good! If you can get the CLA filled out I'll go ahead and merge this :)

**Comment by ohenrik** [comment]
Sounds good :) I have added the contributors agreement. I have also trained a model based on this data: https://github.com/UniversalDependencies/UD_Norwegian-Bokmaal

How do i add a trained model to Spacy?

| Itn. | P.Loss | N.Loss | UAS | NER P. | NER R. | NER F. | Tag % | Token % |
| -- | -- | -- | -- | -- | -- | -- | -- | -- |
| 0 | 500.962 | 0.000 | 83.669 | 0.000 | 0.000 | 0.000 | 93.269 | 100.000 | 3837.8 | 0.0 |
| 1 | 86.554 | 0.000 | 86.384 | 0.000 | 0.000 | 0.000 | 94.396 | 100.000 | 3851.8 | 0.0 |
| 2 | 35.351 | 0.000 | 87.065 | 0.000 | 0.000 | 0.000 | 94.762 | 100.000 | 3712.0 | 0.0 |
| 3 | 21.769 | 0.000 | 87.990 | 0.000 | 0.000 | 0.000 | 94.839 | 100.000 | 3700.4 | 0.0 |
| 4 | 19.490 | 0.000 | 88.264 | 0.000 | 0.000 | 0.000 | 95.020 | 100.000 | 3706.6 | 0.0 |
| 5 | 17.730 | 0.000 | 88.475 | 0.000 | 0.000 | 0.000 | 95.084 | 100.000 | 3750.7 | 0.0 |
| 6 | 16.141 | 0.000 | 88.771 | 0.000 | 0.000 | 0.000 | 95.042 | 100.000 | 3704.5 | 0.0 |
| 7 | 14.906 | 0.000 | 88.720 | 0.000 | 0.000 | 0.000 | 95.139 | 100.000 | 3811.2 | 0.0 |
| 8 | 13.644 | 0.000 | 88.756 | 0.000 | 0.000 | 0.000 | 95.042 | 100.000 | 3806.8 | 0.0 |

**Comment by ohenrik** [comment]
I would also really like to add a noun_chunker to this. However i'm not sure about how to construct the rules for Norwegian. Got any tips on how to find resources for this?

**Comment by ines** [comment]
@ohenrik Thanks for sharing the training results, this looks really good! 👍 

> How do i add a trained model to Spacy?

It'd be really cool to get the model out to the community and have people test it! So if you like, you could publish it on your GitHub profile as a "community model", and we'd be happy to share the link on Twitter to get the word out to people! This would let you keep ownership of the model for now, and make it easier for other Norwegian users to get involved, give feedback and discuss enhancements (for example, tweaking the hyperparameters and adding NER).

You can use the `spacy package` command to package it for spaCy and then run `python setup.py sdist` in the created directory. In the license field of the model's `meta.json`, you can use the license of the corpus (CC BY-SA 4.0, right?). We usually then attach the packaged model to the [GitHub releases](https://github.com/explosion/spacy-models/releases), which works pretty well.

Once the model is tested, we can exchange hyperparameters and include it in our official model build pipeline. The Nordic/Scandinavian languages are definitely next on our list of models we want to train.

**Comment by ohenrik** [comment]
Perfect, i made a model based on iteration 7 (which seemed to have the best result related to the tagger/dependency tree.

| Itn. | P.Loss | N.Loss | UAS | NER P. | NER R. | NER F. | Tag % | Token % |
| -- | -- | -- | -- | -- | -- | -- | -- | -- |
| 7 | 14.906 | 0.000 | 88,72 | 0.000 | 0.000 | 0.000 | 95,139 | 100.000 | 3572.3 | 0.0 |

The repo can be found here: 
https://github.com/ohenrik/nb_core_ud

I wasn't completely sure about how to package and what to include in the repo, so i have included all of it. Please let me know if i should change something.

I trained the model with further iterations and at some point the results plummet, is this just because of overfitting?

| Itn. | P.Loss | N.Loss | UAS | NER P. | NER R. | NER F. | Tag % | Token % |
| -- | -- | -- | -- | -- | -- | -- | -- | -- |
12 | 10.950 | 0.000 | 88,67 | 0.000 | 0.000 | 0.000 | 94,754 | 100.000 | 3909.9 | 0.0
13 | 10.325 | 0.000 | 88,85 | 0.000 | 0.000 | 0.000 | 47,879 | 100.000 | 3673.9 | 0.0

**Comment by ines** [comment]
> I wasn't completely sure about how to package and what to include in the repo, so i have included all of it. Please let me know if i should change something.

Oh cool, I'm surprised it fit into a GitHub repo! You should be able to create a pip-installable zip file like this:

```bash
cd nb_ud_model_seven-0.0.0
python setup.py sdist
```

This will generate a `nb_ud_model_seven-0.0.0.zip` file of in a directory `/dist`. You can then create a new release on GitHub and attach the `.zip` file to it. This file can then be installed as a model package via `pip install`. (This is also how we do it for our official models btw.)

```bash
pip install path/to/nb_ud_model_seven-0.0.0.zip
```

Before we share the model on Twitter, it'd also be cool if the repo had a little README that gives some details and links (the corpus used, the accuracy figures, installation instructions etc.), and maybe a note that it's just experimental / not an official model, so people are not confused by it. Maybe you also want to rename it and give it a more descriptive name, e.g. `nb_dep_ud_sm-0.0.1`?

Anyway, thanks again for your work on this! 👍 Looking forward to sharing this with the Norwegian spaCy community.

> I trained the model with further iterations and at some point the results plummet, is this just because of overfitting?

This is interesting... did the accuracy go up again afterwards or did it stay this low?

**Comment by ohenrik** [comment]
> This is interesting... did the accuracy go up again afterwards or did it stay this low?

It stayed this low, for the remaining iterations.

Strange I though i added this when i used `python setup.py sdist` earlier, I will look into adding the zip file tomorrow. Also I’ll add a description and stats etc :)

**Comment by honnibal** [comment]
One reason accuracy might drop like that is that the batch size is set to increase during training by default, so if the maximum batch size is too large for your problem, maybe the model diverges? The various hyper-parameters are explained here: https://spacy.io/usage/training#section-tips

**Comment by ohenrik** [comment]
Thank you @honnibal I will try that, i though the rules aren't the rules for which nouns to combine where specific to each language. or is the noun chunker also trained somehow? I.e. the rules are determined by the model, not by static rules (as would make sense i guess).

@ines i updated the language model repo. Please take a look and let me know if i should change something. :) (https://github.com/ohenrik/nb_core_ud)

**Comment by ines** [comment]
The noun chunker is made up of rules, based on the dependency parse. There may be language-specific exceptions, or different rules for different label schemes (e.g. German). But if both languages use the same dependency label scheme, the same (or very similar) rules can be used to determine the noun chunks, even if the languages and syntax are pretty different. So as an experiment, you can just try and copy the French `noun_chunks` syntax iterator, and run it over a bunch of Norwegian text. If the dependencies are assigned correctly, the results should be pretty good.

> i updated the language model repo. Please take a look and let me know if i should change something. :) (ohenrik/nb_core_ud)

🎉 One small thing: Do you want to rename the repo to match the updated model name? And should we tweet it out as "contributions / feedback welcome" or something like that?

**Comment by ohenrik** [comment]
@ines no problem, i renamed the repo: https://github.com/ohenrik/nb_dep_ud_sm

And yes, please tweet about it, i would very much like the model to be used and improved :)

**Comment by ines** [comment]
@ohenrik Perfect – will share this later! We definitely have a lot of Norwegian users, so I hope they'll try it out. (Considering there are also quite a few Norwegian Prodigy users now, you might even be able to crowdsource some NER annotations to train the entity recognizer and make it a "core" model!)

Btw, did you manage to get the installable `.zip` file attached or included? It doesn't show up for me, but it'd be cool if people were able to just download one file and run `pip install` to try it out.

**Comment by ohenrik** [comment]
That sounds awesome! I was considering starting a NER annotation project later. I talked to some language scientists at UIO (University in Oslo) earlier, they have a project where they might release the NER data this summer.

@ines I had to strict default python .gitignore settings, so my `dist` folder was excluded again. I have updated it now. https://github.com/ohenrik/nb_dep_ud_sm/tree/master/nb_dep_ud_sm-0.0.1/dist

**Comment by ines** [comment]
@ohenrik Ah, perfect! Maybe you want to add one little section to the README that links to that file and tells users to download the zip and `pip install` it? Just to make sure people don't miss it.

Anyway, we'll definitely tweet about the model later today! 👍

**Comment by ohenrik** [comment]
@ines i have now added and tested install and usage instructions. :)

**Comment by jarib** [comment]
@ohenrik Did you make any progress on NER data from UiO?

**Comment by ohenrik** [comment]
@jarib I did make a version that included a trained NER with accuracy around 85% (if i remember correctly), however this was done for a client, so i have to as if i can share the model. The data behind it should be open source, just need to find the official link, so i worse case you can train the model your self.

**Comment by ohenrik** [comment]
The dataset for NER tagger can be found here:
https://github.com/ltgoslo/norne

I used the files under `ud/nob`

To use it you need to convert it suing the conllubio2json converter (cli tool):
https://github.com/explosion/spaCy/blob/master/spacy/cli/converters/conllubio2json.py

The documentation here can be used (with conllubio2json instead of conllu2json as argument):
https://spacy.io/api/cli#convert

`python -m spacy convert [input_file] [output_dir] --converter=conllubio2json --morphology`

**However, i will publish the model i trained either tomorrow or within the next week.**

**Comment by jarib** [comment]
@ohenrik Great, I actually found the norne data and have already done some training with it.

It looks like NER support was added to the standard `conllu2json` converter in  https://github.com/explosion/spaCy/pull/2594 by @katarkor who is also training a Norwegian model. 

It does look like I'm having the same problem with the model diverging (at the 9th iteration) that you had. Did you experiment with tweaking the batch sizes?

```
Itn.  Dep Loss  NER Loss  UAS     NER P.  NER R.  NER F.  Tag %   Token %  CPU WPS  GPU WPS
0     17340.454 1374.519  84.194  76.510  75.045  75.770  93.720  100.000  4070.4   2743.7
1     421.406   14.045    86.326  77.838  77.977  77.907  94.699  100.000  4122.8   2805.9
2     351.667   9.719     87.010  79.533  79.533  79.533  94.869  100.000  3970.1   2777.2
3     305.092   7.577     87.346  79.621  80.431  80.024  94.954  100.000  4074.4   2795.6
4     272.923   6.345     87.741  79.775  80.491  80.131  95.031  100.000  4107.4   2782.1
5     249.849   5.603     87.846  77.885  77.558  77.721  95.226  100.000  3939.6   2798.2
6     226.618   4.956     87.921  78.375  78.516  78.445  95.102  100.000  3906.4   2764.4
8     192.111   3.775     87.815  77.824  78.336  78.079  95.141  100.000  4086.0   2801.3
9     180.068   3.619     88.053  78.243  78.336  78.289  36.345  100.000  4134.0   2792.8
10    170.211   3.306     88.281  78.406  78.875  78.640  39.170  100.000  4054.2   2799.9
11    160.809   3.233     88.353  78.443  78.396  78.420  45.400  100.000  4028.7   2784.0
12    150.422   2.764     88.216  78.121  78.636  78.378  48.493  100.000  4094.4   2806.8
```

**Comment by jarib** [comment]
@ohenrik I created a new issue where we can continue this – https://github.com/explosion/spaCy/issues/3082
